homework_2/fastq_filtrator.py

```diff
@@ -0,0 +1,173 @@
+        print("ERROR: border value for length input is of invalid type.")
+        print("Please input border value for length using int or float type for onpy upper border or tuple or list type for lower and upper borders.")
+        exit()
+    
+    # check for the valid border values 
+    if not 0 <= min_bound <= max_bound <= 2**32:
+        print("ERROR: border values for length input are invalid.")
+        print("Please note that border values could not be less than 0 and more that 2**32 and that lower border should be before upper border.")
+        exit()
+
+    # checking whether sequence length is appropriate
+    if min_bound <= len(seq) <= max_bound:
+        return True
+    else:
+        return False
+
+
+def quality_check(seq_qual, threshold):
+    """Function for performing read quality check"""
+    # check for the valid quality format
+    if type(threshold) != int and type(threshold) != float:
+        print("ERROR: threshold value input is of invalid type.")
+        print("Please input threshold value using int or float type.")
+        exit()
+    
+    # check for the valid threshold value 
+    if not 0 <= threshold <= 40:
+        print("ERROR: threshold input value is invalid.")
+        print("Please note that threshold value could not be less than 0 and more that 40.")
+        exit()
+
+    # creating variable to sum up the Q-scores
+    q_scores_sum = 0 # here i've decided to sum the Q-scores asap instead of appending Q-scores to the list for memory economy 
+
+    # converting phred33 to Q-scores and summing them up
+    for val in seq_qual:
+        q_scores_sum += (ord(val) - 33) 
+    
+    # checking whether the mean read quality is appropriate
+    if (q_scores_sum / len(seq_qual)) >= threshold:
+        return True
+    else:
+        return False
+
+
+def write_good(reading, file_name):
+    """Function for writing appropriate readings to a file"""
+    # making the filename for readings that passed
+    suffix = "{}_passed.fastq"
+    passed_file_name = suffix.format(file_name)
+
+    # writing the fastq reading to a passed file
+    f = open(passed_file_name, "a")
+    for i in reading:
+        f.write(i + "\n")
+    f.close()
+
+
+def write_bad(reading, file_name):
+    """Function for writing inappropriate readings to a file"""
+    # making the filename for readings that failed
+    suffix = "{}_failed.fastq"
+    failed_file_name = suffix.format(file_name)
+
+    # writing the fastq reading to a failed file
+    f = open(failed_file_name, "a")
+    for i in reading:
+        f.write(i + "\n")
+    f.close()            
+
+"""main function for file reading"""
+def main(
+    input_fastq, 
+    output_file_prefix, 
+    gc_bounds=(0, 100), 
+    length_bounds=(0, 2**32),
+    quality_threshold=0, 
+    save_filtered=False):
+    # file opening and writing each 4 lines to buffer
+    with open(input_fastq, 'r') as inp:
+        buff = []
+        buff_counter = 0
+        for line in inp.readlines():
```

Comment:
Тут хорошая идея с буфером, который мы заполняем по 4 строки, тогда непонятно зачем тут `readlines`, который считывает файл полностью. Его можно смело заменить
```suggestion
        for line in inp:
```